src/cryptographic_primitives/proofs/low_degree_exponent_interpolation.rs

```diff
@@ -0,0 +1,215 @@
+            .iter()
+            .zip(&alpha)
+            .map(|(g, a)| g.clone() * w.evaluate(a))
+            .collect();
+
+        let d = w.degree();
+        let u = Polynomial::<P>::sample(d);
+        let a: Vec<P> = g
+            .iter()
+            .zip(&alpha)
+            .map(|(g, a)| g.clone() * u.evaluate(a))
+            .collect();
+
+        let hash_input: Vec<&P> = g.iter().chain(&x).chain(&a).collect();
+        let e = H::create_hash_from_ge::<P>(hash_input.as_slice());
+
+        let z = &u - &(w * &e);
+
+        let statement = LdeiStatement {
+            alpha,
+            g,
+            x,
+            degree: d,
+        };
+        let proof = LdeiProof { a, e, z };
+
+        Ok((statement, proof))
+    }
+
+    /// Verifies correctness of a statement
+    ///
+    /// ## Protocol
+    ///
+    /// The verifier checks that `e = H(g1,...,gm,x1,...,xm,a1,...,am)`, that
+    /// `deg(z) ≤ d`, and that `a_i = g_i^z(αlpha_i) * x_i^e` for all i, and accepts if all of this is
+    /// true, otherwise rejects.
+    pub fn verify<H>(&self, statement: &LdeiStatement<P>) -> Result<(), ProofError>
+    where
+        H: Hash,
+    {
+        let hash_input: Vec<&P> = statement
+            .g
+            .iter()
+            .chain(&statement.x)
+            .chain(&self.a)
+            .collect();
+        let e = H::create_hash_from_ge::<P>(hash_input.as_slice());
+        if e != self.e {
+            return Err(ProofError);
+        }
+        if self.z.degree() > statement.degree {
+            return Err(ProofError);
+        }
+
+        let expected_a: Vec<_> = statement
+            .g
+            .iter()
+            .zip(&statement.alpha)
+            .zip(&statement.x)
+            .map(|((g, a), x)| g.clone() * self.z.evaluate(&a) + x.clone() * e.clone())
+            .collect();
+
+        if self.a == expected_a {
+            Ok(())
+        } else {
+            Err(ProofError)
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum LdeiProveError {
+    #[error("`alpha`s are not pairwise distinct")]
+    AlphaNotPairwiseDistinct,
+    #[error("alpha.len() != g.len()")]
+    AlphaLengthDoesntMatchG,
+}
+
+#[cfg(test)]
+mod tests {
+    use std::iter;
+
+    use crate::arithmetic::BigInt;
+    use crate::cryptographic_primitives::hashing::hash_sha256::HSha256;
+    use crate::elliptic::curves::traits::ECScalar;
+    use crate::test_for_all_curves;
+
+    use super::*;
+
+    test_for_all_curves!(correctly_proofs);
+    fn correctly_proofs<P>()
+    where
+        P: ECPoint + Clone + PartialEq,
+        P::Scalar: ECScalar + Clone + PartialEq,
+    {
+        let poly = Polynomial::<P>::sample(5);
+
+        let alpha: Vec<P::Scalar> = (1..=10).map(|i| ECScalar::from(&BigInt::from(i))).collect();
+        let g: Vec<P> = iter::repeat_with(ECScalar::new_random)
+            .map(|x| P::generator() * x)
+            .take(10)
+            .collect();
+
+        let (statement, proof) =
```

Comment:
the prover should get a statement and a witness.

Reply:
Done. Can you confirm that it now looks like you expected?